### DISCUSSION TITLE
It will be awesome to use it on different board as well

### DISCUSSION THREAD

**Discussion by niradler** [toitlang/toit#17] [Ideas]
I guess its matter of focus but I suggest you don't limit the language to esp32.
💡 Just an idea to think on 💡

**Comment by kasperl** [comment]
Thanks! It is definitely worth considering. Do you have suggestions on interesting boards or CPUs to support?

**Reply by niradler** [reply]
not sure im up to date, but esp32 is not the only board out there, and companies rarely use it in consumer electronics.
beside this I prefer to write toit then c++/arduino so even this can be cool Atmel have really mature tooling.

**Reply by mikini** [reply]
A straightforward path could also be to add ESP8266 support. 

That would open up use on a lot of existing products based on this which is already out there (like fx. [Sonoff](https://sonoff.tech/)/[itead](https://itead.cc/)/[Nextion](https://nextion.tech/datasheets/)) and all sorts of derivates that projects [like Tasmota also supports](https://tasmota.github.io/docs/Supported-Modules/).

ESP8266 is not directly supported by the [ESP-IDF](https://docs.espressif.com/projects/esp-idf/en/latest/) currently used by Toit, but I guess it would be a lot less work than supporting a completely different architecture such as AVR (Arduino etc.) or STM32 (blue pill/red pill).

**Reply by dsobotta** [reply]
I started an experimental fork for supporting Toit on sifive-unmatched risc-v boards ( https://github.com/dsobotta/toit-riscv ).  Still working through some compiler errors, but the  changes and updated documentation should save a few hours of debugging.  Once the vm is running properly, I'll start working through a pull request for merging back upstream.

In the meantime, I'd be happy to incorporate any PRs that bring us closer to that goal.

**Reply by kasperl** [reply]
@dsobotta Do let me know if you need any help with the port!

**Reply by justind000** [reply]
I would imagine it would take considerable effort to port this to a non-ESP32 board given the reliance on ESP-IDF. 

I'd like to see it on the other ESP32 variants though. The original ESP32's have had two or three newer versions released in the last couple years. 

ESP8266's are great and marginally less expensive, but relatively ancient when compared to the newer Espressif options.

**Reply by dsobotta** [reply]
@justind000 Now that I have Toit on RISC-V/Linux running, I intend to work on Toit for RISC-V/Embedded.  I'll probably start with evaluating the [ESP32-C3](https://hackaday.com/2021/02/08/hands-on-the-risc-v-esp32-c3-will-be-your-new-esp8266/), as support is already available in ESP-IDF.  

Since ESP-IDF is built atop a fork of FreeRTOS, and the differences are [somewhat localized and mostly additions](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/freertos-smp.html), it should be viable to eventually target most modern microcontrollers that already have FreeRTOS support.  The biggest issue I see is determining which chips are worth supporting and maintaining merges from both ESP-IDF and vanilla FreeRTOS.

That said, as far as tiny Linux devices go, the Raspberry Pi Zero 2W looks like a really great candidate for just $15...  Given how smooth the RISC-V/Linux port went, I suspect it would take a couple days or less to support ARM/Linux.  I'll probably take an initial look at it in the next week or two.

**Comment by mess-maker** [comment]
hi,
the atmel SAMD21/SAMD51 are a popular family with nice product like the [wio terminal](https://www.gotronic.fr/art-wio-terminal-102991299-31802.htm) or some adafruit feather

the RP2040 family with the raspberry pi pico

the NRF51 (notably micro:bit) and NRF52 (micro:bit v2 and some feather)

the STM32 family ([weact studio black pill](https://fr.aliexpress.com/item/1005001456186625.html?pdp_ext_f=%7B%22sku_id%22:%2212000016817645221%22,%22ship_from%22:%22CN%22%7D&gps-id=pcStoreLeaderboard&scm=1007.22922.122102.0&scm_id=1007.22922.122102.0&scm-url=1007.22922.122102.0&pvid=faae8e2b-c48a-46f6-a74e-11c4074269ee&spm=a2g0o.store_pc_home.smartLeaderboard_6000147819209.1005001456186625))

thoose should cover a big chunk of the ''popular'' microcontrollers

**Reply by dsobotta** [reply]
I'd love to see support for the RP2040; it's a joy to work with and expanding rapidly in popularity.

FWIW, it is compatible with vanilla RTOS; so maybe it's not much of a stretch to get preliminary support up and running.

**Reply by dsobotta** [reply]
[UDOO Key](https://www.kickstarter.com/projects/udoo/udoo-key-the-4-ai-platform) is shipping by end of Jan for those that Kickstarted.  The board contains both ESP32 and RP2040; so it could prove to be a good starting point for adding general RP2040 support.

I'll probably start tinkering with mine early Feb.  For now my focus is still on ESP32-C3.

**Comment by skye0402** [comment]
As I wrote [in a different thread](https://github.com/toitlang/toit/discussions/236) I propose the [Arduino Nano 33 BLE Sense](https://store-usa.arduino.cc/products/arduino-nano-33-ble-sense) as a board to try to move Toit to. Reason being that it comes with sensors and BLE on board and was widely use for TinyML with TensorFlow.
I also agree that the RP2040 would be a very interesting board to have Toit running on as the community is big. The Nano 33 has an Arm Cortex M4, the RP2040 two Arm Cortex-M0+.

**Reply by mess-maker** [reply]
the nano 33 family have nRF52840 not cortex M4

**Reply by skye0402** [reply]
@mess-maker the [nRF52840](https://infocenter.nordicsemi.com/pdf/nRF52840_PS_v1.1.pdf) uses Cortex-M4. Please check.

**Comment by mess-maker** [comment]
if you want RP2040 and wifi on board you also have the new [nano RP2040 connect](https://docs.arduino.cc/hardware/nano-rp2040-connect)
